lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp

```diff
@@ -0,0 +1,1070 @@
+//                             Block State 
+//===----------------------------------------------------------------------===//
+
+struct BlockState {
+  /// A bit vector for which the ith bit represents the ith refcounted root in
+  /// RCRootVault.
+  ///
+  /// NOTE: we could do the data flow with BBSetIn or BBSetOut, but that would
+  /// require us to create a temporary copy to check whether the BBSet has
+  /// changed after the genset and killset has been applied.
+  llvm::SmallBitVector BBSetIn;
+
+  /// A bit vector for which the ith bit represents the ith refcounted root in
+  /// RCRootVault.
+  llvm::SmallBitVector BBSetOut;
+
+  /// A bit vector for which the ith bit represents the ith refcounted root in
+  /// RCRootVault. If the bit is set, that means this basic block creates a
+  /// retain which can be sunk or a release which can be hoisted.
+  llvm::SmallBitVector BBGenSet;
+
+  /// A bit vector for which the ith bit represents the ith refcounted root in
+  /// RCRootVault. If this bit is set, that means this basic block stops retain
+  /// or release of the refcounted root to be moved across.
+  llvm::SmallBitVector BBKillSet;
+
+  /// A bit vector for which the ith bit represents the ith refcounted root in
+  /// RCRootVault. If this bit is set, that means this is potentially a retain
+  /// or release that can be sunk or hoisted to this point. This is used to
+  /// optimize the time for computing genset and killset.
+  /// NOTE: this vector contains an approximation of whether there will be a
+  /// retain or release to a certain point of a basic block.
+  llvm::SmallBitVector BBMaxSet;
+};
+
+/// CodeMotionContext - This is the base class which retain code motion and
+/// release code motion inherits from. It defines an interface as to how the
+/// code motion precedure should be.
+class CodeMotionContext {
+protected:
+  /// Dataflow needs multiple iteration to converge. If this is false, then we
+  /// do not need to generate the genset or killset, i.e. we can simply do 1
+  /// pessimistic data flow iteration.
+  bool MultiIteration;
+
+  /// The allocator we are currently using.
+  llvm::BumpPtrAllocator &BPA;
+
+  /// Current function we are analyzing.
+  SILFunction *F;
+
+  /// Current post-order we are using.
+  PostOrderFunctionInfo *PO;
+
+  /// Current alias analysis we are using.
+  AliasAnalysis *AA;
+
+  /// Current rc-identity we are using.
+  RCIdentityFunctionInfo *RCFI;
+
+  /// All the unique refcount roots retained or released in the function.
+  llvm::SetVector<SILValue> RCRootVault;
+
+  /// Contains a map between rc roots to their index in the RCRootVault.
+  /// used to facilitate fast rc roots to index lookup.
+  llvm::DenseMap<SILValue, unsigned> RCRootIndex;
+
+  /// All the retains or releases originally in the function. Eventually
+  /// they will all be removed after all the new ones are generated.
+  llvm::DenseSet<SILInstruction *> RCInstructions;
```

Comment:
You should consider SmallPtrSet instead of DenseSet. I believe DenseSet, like DenseMap, starts with size 64 which is going to be much larger than what you need in most cases.